File: app/src/main/java/io/mgba/DI/Modules/ModelModule.java
package io.mgba.DI.Modules;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.mgba.Data.Remote.Interfaces.IRequest;
import io.mgba.Data.Remote.RetrofitClient;
import io.mgba.Model.IO.FilesManager;
import io.mgba.Model.IO.LocalDB;
import io.mgba.Model.Interfaces.IDatabase;
import io.mgba.Model.Interfaces.IFilesManager;
import io.mgba.Model.Interfaces.ILibrary;
import io.mgba.Model.Library;
import io.mgba.Utils.IDeviceManager;
import io.mgba.mgba;

@Module
public class ModelModule {
    private final IDeviceManager manager;
    private mgba application;
    private String path;

    public ModelModule(mgba application, String path, IDeviceManager manager) {
        this.application = application;
        this.path = path;
        this.manager = manager;
    }

    @Provides
    @Singleton
    public ILibrary provideLibrary(){
        return new Library(application);
    }

    @Provides
    @Singleton
    public IDatabase provideDatabase(){
        return new LocalDB(application);
    }

    @Provides
    @Singleton
    public IFilesManager provideFileManager(){
        return new FilesManager(path);
    }

    @Provides
    public IDeviceManager provideDeviceManager(){
        return manager;
    }

    @Provides
    public IRequest provideRetrofitClient(){
        final RetrofitClient retrofitClient = new RetrofitClient();
        return retrofitClient.getClient(IRequest.BASE_URL).create(IRequest.class);
    }

}

File: app/src/main/java/io/mgba/di/modules/ModelModule.kt
package io.mgba.di.modules

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import io.mgba.data.remote.interfaces.IRequest
import io.mgba.data.remote.RetrofitClient
import io.mgba.model.io.FilesManager
import io.mgba.model.io.LocalDB
import io.mgba.model.interfaces.IDatabase
import io.mgba.model.interfaces.IFilesManager
import io.mgba.model.interfaces.ILibrary
import io.mgba.model.Library
import io.mgba.utilities.IDeviceManager
import io.mgba.utilities.IResourcesManager
import io.mgba.mgba

@Module
class ModelModule(private val application: mgba, private val path: String) {

    @Provides
    @Singleton
    fun provideLibrary(): ILibrary {
        return Library(application)
    }

    @Provides
    @Singleton
    fun provideDatabase(): IDatabase {
        return LocalDB(application)
    }

    @Provides
    @Singleton
    fun provideFileManager(): IFilesManager {
        return FilesManager(path)
    }

    @Provides
    fun provideDeviceManager(): IDeviceManager {
        return application
    }

    @Provides
    fun provideResourceManager(): IResourcesManager {
        return application
    }


    @Provides
    fun provideRetrofitClient(): IRequest {
        val retrofitClient = RetrofitClient()
        return retrofitClient.getClient(IRequest.BASE_URL).create(IRequest::class.java!!)
    }

}
